cl_manycore/libraries/bsg_manycore_eva.cpp

```diff
@@ -354,11 +354,254 @@ int default_eva_to_npa(const hb_mc_config_t *cfg,
+                hb_mc_coordinate_get_y(host) == hb_mc_npa_get_y(npa);
+}
+
+/**
+ * Check if an NPA is a local address.
+ * @param[in] config  An initialized manycore configuration struct
+ * @param[in] npa     An npa to translate
+ * @param[in] tgt     Coordinates of the target tile
+ * @return true if the NPA is local, false otherwise.
+ */
+static bool default_npa_is_local(const hb_mc_config_t *config,
+                                 const hb_mc_npa_t *npa,
+                                 const hb_mc_coordinate_t *tgt)
+{
+        // does your coordinate map to this tgt v-core and is your epa valid?
+        return (hb_mc_npa_get_x(npa) == hb_mc_coordinate_get_x(*tgt)) &&
+                (hb_mc_npa_get_y(npa) == hb_mc_coordinate_get_y(*tgt)) &&
+                default_local_epa_is_valid(config, hb_mc_npa_get_epa(npa), tgt);
+}
+
+/**
+ * Check if an NPA is a global address.
+ * @param[in] config  An initialized manycore configuration struct
+ * @param[in] npa     An npa to translate
+ * @param[in] tgt     Coordinates of the target tile
+ * @return true if the NPA is global, false otherwise.
+ */
+static bool default_npa_is_global(const hb_mc_config_t *config,
+                                  const hb_mc_npa_t *npa,
+                                  const hb_mc_coordinate_t *tgt)
+{
+        hb_mc_idx_t base_x = 0;
+        hb_mc_idx_t base_y = 0;
+        hb_mc_idx_t ceil_x = hb_mc_coordinate_get_x(hb_mc_config_get_dimension(config));
+        hb_mc_idx_t ceil_y = hb_mc_coordinate_get_y(hb_mc_config_get_dimension(config));
+
+        // does your coordinate map to any v-core and is your epa valid?
+        return (hb_mc_npa_get_x(npa) >= base_x) && (hb_mc_npa_get_x(npa) <= ceil_x) &&
+                (hb_mc_npa_get_y(npa) >= base_y) && (hb_mc_npa_get_y(npa) <= ceil_y) &&
+                default_local_epa_is_valid(config, hb_mc_npa_get_epa(npa), tgt);
+
+}
+
+/**
+ * Translate a global NPA to an EVA.
+ * @param[in]  cfg      An initialized manycore configuration struct
+ * @param[in]  origin   Coordinate of the origin for this tile's group
+ * @param[in]  tgt      Coordinates of the target tile
+ * @param[in]  npa      An npa to translate
+ * @param[out] eva      An eva to set by translating #npa
+ * @param[out] sz       The size in bytes of the EVA segment for the #npa
+ * @return HB_MC_SUCCESS if succesful. HB_MC_FAIL otherwise.
+ */
+static int default_npa_to_eva_dram(const hb_mc_config_t *cfg,
+                                   const hb_mc_coordinate_t *origin,
+                                   const hb_mc_coordinate_t *tgt,
+                                   const hb_mc_npa_t *npa,
+                                   hb_mc_eva_t *eva,
+                                   size_t *sz)
+{
+        // build the eva
+        hb_mc_eva_t addr = 0;
+        addr |= hb_mc_npa_get_epa(npa); // set the byte address
+        addr |= hb_mc_npa_get_x(npa) << DEFAULT_GLOBAL_X_BITIDX; // set the x coordinate
```

Comment:
This isn't correct - the X index of DRAM depends on the network address width. DEFAULT_GLOBAL is for accesses to the Local Memory of other cores in the array

Reply:
See default_eva_to_npa_dram - where shift and mask is computed.